src/OpenTelemetry/Trace/Configuration/TracerFactory.cs

```diff
@@ -0,0 +1,265 @@
+        {
+            if (this.isBuilt)
+            {
+                throw new InvalidOperationException(nameof(TracerBuilder) + "is already built");
+            }
+
+            this.binaryFormat = binaryFormat;
+            return this;
+        }
+
+        internal ITracer Build(Resource resource)
+        {
+            ITracer tracer;
+            if (!this.isBuilt)
+            {
+                if (this.tracerConfigurationOptions == null)
+                {
+                    // TODO separate sampler from options
+                    this.tracerConfigurationOptions = new TracerConfigurationOptions(this.sampler);
+                }
+
+                if (this.spanExporter == null)
+                {
+                    // TODO log warning
+                    this.spanExporter = new NoopSpanExporter();
+                }
+
+                this.spanProcessor = this.processorFactory != null
+                    ? this.processorFactory(this.spanExporter)
+                    : new BatchingSpanProcessor(this.spanExporter);
+
+                tracer = new Tracer(
+                    this.spanProcessor,
+                    this.tracerConfigurationOptions,
+                    this.binaryFormat,
+                    this.textFormat,
+                    resource);
+
+                if (this.collectorFactories != null)
+                {
+                    foreach (var collector in this.collectorFactories)
+                    {
+                        var collectorInstance = collector.Invoke(tracer);
+                        if (collectorInstance is IDisposable disposableCollector)
+                        {
+                            this.disposables.Add(disposableCollector);
+                        }
+                    }
+                }
+
+                this.isBuilt = true;
+            }
+            else
+            {
+                tracer = new Tracer(
+                    this.spanProcessor,
+                    this.tracerConfigurationOptions,
+                    this.binaryFormat,
+                    this.textFormat,
+                    resource);
+            }
+
+            return tracer;
+        }
+
+        public void Dispose()
+        {
+            if (this.spanProcessor is IDisposable disposableProcessor)
+            {
+                disposableProcessor.Dispose();
+            }
+
+            foreach (var disposable in this.disposables)
+            {
+                disposable.Dispose();
+            }
+        }
+    }
+
+    public class TracerFactory : TracerFactoryBase, IDisposable
+    {
+        private readonly object lck = new object();
+        private readonly Dictionary<TracerRegistryKey, ITracer> tracerRegistry = new Dictionary<TracerRegistryKey, ITracer>();
+        private ITracer defaultTracer;
+        private readonly TracerBuilder builder;
+
+        private TracerFactory(TracerBuilder builder)
+        {
+            this.builder = builder;
```

Comment:
Is this a temporary?

I'd copy the values off trace builder here instead of holding onto it. This feels like odd coupling. And by copying the values you don't need to worry about someone modifying the builder later.